### PR TITLE
Serge/feat/get at index

### PR DESCRIPTION
**Pull request by serge-21**

### CODE REVIEW

**Comment by williambl** [comment]
#53 is now merged, this can make use of that

**Comment by serge-21** [comment]
thank you for the heads up will do!!! :)

File: basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/GetAtIndexNodeType.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    public Set<Input<?>> inputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        return Set.of(
+            new NodeConnector.Input<>(nodeId, "input", ListDataType.GENERIC_LIST),
+            new NodeConnector.Input<>(nodeId, "index", DataType.NUMBER)
+        );
+    }
+
+    @Override
+    public Set<Output<?>> outputs(UUID nodeId, Map<String, DataBox<?>> settings, Map<String, DataType<?>> inputTypes) {
+        var inputType = inputTypes.getOrDefault("input_type", OptionalDataType.GENERIC_OPTIONAL);
```
Comment (williambl) [inline basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/GetAtIndexNodeType.java:34]: ? There is no input called 'input_type' or one that takes an optional - this won't work.

File: basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+        Map<String, DataType<?>> inputTypes = Map.of(
+            "input_type", DataType.STRING
+        );
+
+        var output = BasicNodes.GET_AT_INDEX.compute(input, Map.of(), inputTypes);
+        Assertions.assertEquals("i0", output.get("output").join().value());
+        Assertions.assertEquals(output.get("output").join().type(), DataType.STRING);
+    }
+
```
Comment (williambl) [inline basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java:140]: There should be a test here that involves connecting the node in a graph, to test the dynamic typing.

**Review by williambl — CHANGES_REQUESTED** [review]

File: basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java (hunk 2 above)
Comment (serge-21) [reply basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java:140]: Currently, there is no node with an output of a list. How would you suggest doing this test?

**Review by ShaksterNano — APPROVED** [review]

**Comment by fenn7** [comment]
address changes needed

**Comment by serge-21** [comment]
@fenn7 I still do not know how to do this since there is really no node with an output of a list type.
also is this the correct file to test it or should this be in the graph file test since this is a quality of a graph rather than two nodes 🤔

@williambl ops on this?

File: basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        Map<String, DataBox<?>> input = Map.of(
+            "input", ListDataType.GENERIC_LIST.create(testArrayList),
+            "index", DataType.NUMBER.create(0.0)
+        );
+
+        Map<String, DataType<?>> inputTypes = Map.of(
+            "input", DataType.NUMBER
+        );
+
+        var output = BasicNodes.GET_AT_INDEX.compute(input, Map.of(), inputTypes);
```
Comment (williambl) [inline basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java:101]:
```suggestion
		var node = Node.create(BasicNodes.GET_AT_INDEX)
			.updateInputTypes(inputTypes);
        Assertions.assertEquals(DataType.NUMBER, node.getOutputWithName("output").orElseThrow().type());
        var output = BasicNodes.GET_AT_INDEX.compute(input, Map.of(), node.inputTypes());
```

File: basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java (hunk 2 above)
Comment (williambl) [reply basic_nodes/src/test/java/club/mondaylunch/gatos/basicnodes/test/GetAtIndexNodeTest.java:143]: Uh yeah IDK what I was thinking there. See the suggestion above, that's the only thing I think that's missing from the test.

File: basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/GetAtIndexNodeType.java (hunk 1 above)
Comment (serge-21) [reply basic_nodes/src/main/java/club/mondaylunch/gatos/basicnodes/GetAtIndexNodeType.java]: this has been resolved

**Review by williambl — APPROVED** [review]

**Pull request merged**